x/tokenomics/keeper/token_logic_modules.go

```diff
@@ -0,0 +1,443 @@
+// tokenomics module account to the specified address.
+func (k Keeper) sendRewardsToAccount(
+	ctx context.Context,
+	addr string,
+	settlementAmtFloat *big.Float,
+	allocation float64,
+) error {
+	logger := k.Logger().With("method", "mintRewardsToAccount")
+
+	accountAddr, err := cosmostypes.AccAddressFromBech32(addr)
+	if err != nil {
+		return err
+	}
+
+	coinsToAccAmt, _ := new(big.Float).Mul(settlementAmtFloat, big.NewFloat(allocation)).Int64()
+	coinsToAcc := sdk.NewCoins(cosmostypes.NewCoin(volatile.DenomuPOKT, math.NewInt(coinsToAccAmt)))
+	if err := k.bankKeeper.SendCoinsFromModuleToAccount(
+		ctx, suppliertypes.ModuleName, accountAddr, coinsToAcc,
+	); err != nil {
+		return err
+	}
+	logger.Info(fmt.Sprintf("sent (%v) coins from the tokenomics module to the account with address %q", coinsToAcc, addr))
+
+	return nil
+}
+
+func (k Keeper) handleOverservicedApplication(
+	ctx context.Context,
+	application *apptypes.Application,
+	settlementCoins cosmostypes.Coin,
+) (
+	newSettlementCoins cosmostypes.Coin,
+	err error,
+) {
+	logger := k.Logger().With("method", "handleOverservicedApplication")
+	// over-serviced application
+	logger.Warn(fmt.Sprintf(
+		"THIS SHOULD NEVER HAPPEN. Application with address %s needs to be charged more than it has staked: %v > %v",
+		application.Address,
+		settlementCoins,
+		application.Stake,
+	))
+
+	// TODO_MAINNET(@Olshansk, @RawthiL): The application was over-serviced in the last session so it basically
+	// goes "into debt". Need to design a way to handle this when we implement
+	// probabilistic proofs and add all the parameter logic. Do we touch the application balance?
+	// Do we just let it go into debt? Do we penalize the application? Do we unstake it? Etc...
+	expectedBurn := settlementCoins
+
+	applicationOverservicedEvent := &tokenomicstypes.EventApplicationOverserviced{
+		ApplicationAddr: application.Address,
+		ExpectedBurn:    &expectedBurn,
+		EffectiveBurn:   application.GetStake(),
+	}
+	eventManager := cosmostypes.UnwrapSDKContext(ctx).EventManager()
+	if err := eventManager.EmitTypedEvent(applicationOverservicedEvent); err != nil {
+		return cosmostypes.Coin{}, tokenomicstypes.ErrTokenomicsApplicationOverserviced.Wrapf(
+			"application address: %s; expected burn %s; effective burn: %s",
+			application.GetAddress(),
+			expectedBurn.String(),
+			application.GetStake().String(),
+		)
+	}
+	return *application.Stake, nil
+}
+
+// computeUnitsToCoin calculates the amount of uPOKT to mint based on the number of compute units.
+func (k Keeper) computeUnitsToCoin(ctx context.Context, numComputeUnits uint64, service *sharedtypes.Service) (cosmostypes.Coin, error) {
+	computeUnitsPerRelay := service.ComputeUnitsPerRelay
+	computeUnitsToTokensMultiplier := k.GetParams(ctx).ComputeUnitsToTokensMultiplier
+	upoktAmount := math.NewInt(int64(numComputeUnits * computeUnitsPerRelay * computeUnitsToTokensMultiplier))
```

Comment:
:thinking: This is intriguing and may be an indicator that we're doing something wrong.

Given the variable name and purpose of `computeUnitsPerRelay`, I'm expecting it to be multiplied by `numRelays` instead of `numComputeUnits`.

We are currently using the `numComputeUnits` we get from `root.Sum`, which effectively returns the compute units sum, but I think we should:
* Either use `root.Count` which returns the number of relays and do `relayCount * computeUnitsPerRelay * computeUnitsToTokenMultiplier`.
* Or use `root.Sum` and do `numComuteUnits * computeUnitsToTokenMultiplier`.

Doing a little bit of investigation; and think this is currently working fine because:
* Off-chain, the `SessionManager` is updating the `SMST` with `smst.Update(hash, bytes, 1)` which makes `root.Sum` and `root.Count` equal.
* On-chain, the `Proof` module is not verifying that `root.Sum() == root.Count() * service.ComputeUnitsPerRelay`

Reply:
Thank you for catching this!!!!

I kept it a little simpler with an explanation in the comments. Let me know what you think.